File: src/go/core/parity_test.go
package dicebear

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	"github.com/dicebear/dicebear-go/v10/internal/initials"
	"github.com/dicebear/dicebear-go/v10/internal/num"
	"github.com/dicebear/dicebear-go/v10/internal/prng"
)

// Cross-language parity: assert the primitives produce exactly the values in the
// shared fixtures under <repo>/tests/fixtures/parity/, the same the JS, PHP,
// Python and Rust suites run against. The tests skip gracefully when the
// fixtures are absent (e.g. in the split dicebear-go repo).

func fixturePath(t *testing.T, name string) string {
	t.Helper()
	p := filepath.Join("..", "..", "..", "tests", "fixtures", "parity", name)
	if _, err := os.Stat(p); err != nil {
		t.Skipf("parity fixtures not available (%s)", p)
	}
	return p
}

func readFixture(t *testing.T, name string, v any) {
	t.Helper()
	b, err := os.ReadFile(fixturePath(t, name))
	if err != nil {
		t.Fatalf("read %s: %v", name, err)
	}
	if err := json.Unmarshal(b, v); err != nil {
		t.Fatalf("parse %s: %v", name, err)
	}
}

func strID(s string) string { return s }

func TestFnv1aParity(t *testing.T) {
	var cases []struct {
		Input string `json:"input"`
		Hash  uint32 `json:"hash"`
		Hex   string `json:"hex"`
	}
	readFixture(t, "fnv1a.json", &cases)

	for _, c := range cases {
		if got := prng.Hash(c.Input); got != c.Hash {
			t.Errorf("Hash(%q) = %d, want %d", c.Input, got, c.Hash)
		}
		if got := prng.Hex(c.Input); got != c.Hex {
			t.Errorf("Hex(%q) = %s, want %s", c.Input, got, c.Hex)
		}
	}
}

func TestMulberry32Parity(t *testing.T) {
	var cases []struct {
		Seed     uint32 `json:"seed"`
		Sequence []struct {
			Float float64 `json:"float"`
			State int32   `json:"state"`
		} `json:"sequence"`
	}
	readFixture(t, "mulberry32.json", &cases)

	for _, c := range cases {
		m := prng.NewMulberry32(c.Seed)
		for i, step := range c.Sequence {
			f := m.NextFloat()
			s := m.SignedState()
			if f != step.Float {
				t.Errorf("seed %d step %d: float = %v, want %v", c.Seed, i, f, step.Float)
			}
			if s != step.State {
				t.Errorf("seed %d step %d: state = %d, want %d", c.Seed, i, s, step.State)
			}
		}
	}
}

type rangeFixture struct {
	Min  float64  `json:"min"`
	Max  float64  `json:"max"`
	Step *float64 `json:"step"`
}

func (r rangeFixture) toRange() *prng.Range {
	return &prng.Range{Min: r.Min, Max: r.Max, Step: r.Step}
}

func TestPrngParity(t *testing.T) {
	var f struct {
		GetValue []struct {
			Seed, Key string
			Result    float64
		} `json:"getValue"`
		Pick []struct {
			Seed, Key string
			Items     []string
			Result    *string
		} `json:"pick"`
		WeightedPick []struct {
			Seed, Key string
			Weights   map[string]float64
			Result    *string
		} `json:"weightedPick"`
		Bool []struct {
			Seed, Key  string
			Likelihood float64
			Result     bool
		} `json:"bool"`
		Float []struct {
			Seed, Key string
			Range     rangeFixture
			Result    float64
		} `json:"float"`
		Integer []struct {
			Seed, Key string
			Range     rangeFixture
			Result    int
		} `json:"integer"`
		Shuffle []struct {
			Seed, Key string
			Items     []string
			Result    []string
		} `json:"shuffle"`
	}
	readFixture(t, "prng.json", &f)

	for _, c := range f.GetValue {
		if got := prng.New(c.Seed).GetValue(c.Key); got != c.Result {
			t.Errorf("GetValue(%q, %q) = %v, want %v", c.Seed, c.Key, got, c.Result)
		}
	}

	for _, c := range f.Pick {
		got, ok := prng.Pick(prng.New(c.Seed), c.Key, c.Items, strID)
		if !matchOptional(got, ok, c.Result) {
			t.Errorf("Pick(%q, %q, %v) = (%q, %v), want %v", c.Seed, c.Key, c.Items, got, ok, c.Result)
		}
	}

	for _, c := range f.WeightedPick {
		got, ok := prng.New(c.Seed).WeightedPick(c.Key, c.Weights)
		if !matchOptional(got, ok, c.Result) {
			t.Errorf("WeightedPick(%q, %q, %v) = (%q, %v), want %v", c.Seed, c.Key, c.Weights, got, ok, c.Result)
		}
	}

	for _, c := range f.Bool {
		if got := prng.New(c.Seed).Bool(c.Key, c.Likelihood); got != c.Result {
			t.Errorf("Bool(%q, %q, %v) = %v, want %v", c.Seed, c.Key, c.Likelihood, got, c.Result)
		}
	}

	for _, c := range f.Float {
		if got := prng.New(c.Seed).Float(c.Key, c.Range.toRange()); got != c.Result {
			t.Errorf("Float(%q, %q) = %v, want %v", c.Seed, c.Key, got, c.Result)
		}
	}

	for _, c := range f.Integer {
		if got := prng.New(c.Seed).Integer(c.Key, c.Range.toRange()); got != c.Result {
			t.Errorf("Integer(%q, %q) = %d, want %d", c.Seed, c.Key, got, c.Result)
		}
	}

	for _, c := range f.Shuffle {
		got := prng.New(c.Seed).Shuffle(c.Key, c.Items)
		if !equalStrings(got, c.Result) {
			t.Errorf("Shuffle(%q, %q, %v) = %v, want %v", c.Seed, c.Key, c.Items, got, c.Result)
		}
	}
}

func TestNumberParity(t *testing.T) {
	var cases []struct {
		Input  float64 `json:"input"`
		Output string  `json:"output"`
	}
	readFixture(t, "numbers.json", &cases)

	for _, c := range cases {
		if got := num.Format(c.Input); got != c.Output {
			t.Errorf("num.Format(%v) = %q, want %q", c.Input, got, c.Output)
		}
	}
}

func TestInitialsParity(t *testing.T) {
	var cases []struct {
		Seed   string `json:"seed"`
		Result string `json:"result"`
	}
	readFixture(t, "initials.json", &cases)

	for _, c := range cases {
		if got := initials.FromSeed(c.Seed); got != c.Result {
			t.Errorf("initials.FromSeed(%q) = %q, want %q", c.Seed, got, c.Result)
		}
	}
}

func matchOptional(got string, ok bool, want *string) bool {
	if want == nil {
		return !ok
	}
	return ok && got == *want
}

func equalStrings(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

File: src/php/core/tests/ParityTest.php
<?php

declare(strict_types=1);

namespace DiceBear\Tests;

use DiceBear\Avatar;
use DiceBear\Error\CircularColorReferenceError;
use DiceBear\Error\ValidationError;
use DiceBear\OptionsDescriptor;
use DiceBear\Prng;
use DiceBear\Prng\Fnv1a;
use DiceBear\Prng\Mulberry32;
use DiceBear\Style;
use DiceBear\Utils\Color;
use DiceBear\Utils\Initials;
use DiceBear\Utils\Number;
use PHPUnit\Framework\Attributes\DataProvider;
use PHPUnit\Framework\TestCase;

// Cross-language parity tests. Reads the shared fixtures at
// tests/fixtures/parity/ and asserts that the PHP implementation produces
// exactly the values committed there. The JS test suite at
// src/js/core/tests/Parity.test.js reads the same fixtures, so any
// divergence between the two implementations will surface as a failure
// on whichever side drifts.
//
// To regenerate the fixtures: `npm run fixtures:parity` (from repo root).
class ParityTest extends TestCase
{
    private const FIXTURES_DIR = __DIR__ . '/../../../../tests/fixtures/parity';

    /** @return array<string, mixed> */
    private static function loadFixture(string $relPath): array
    {
        $raw = file_get_contents(self::FIXTURES_DIR . '/' . $relPath);
        return json_decode($raw, true, flags: JSON_THROW_ON_ERROR);
    }

    // -----------------------------------------------------------------------
    // Fnv1a
    // -----------------------------------------------------------------------

    public static function fnv1aProvider(): array
    {
        $cases = [];
        foreach (self::loadFixture('fnv1a.json') as $entry) {
            $cases['hash ' . json_encode($entry['input'])] = [$entry];
        }
        return $cases;
    }

    #[DataProvider('fnv1aProvider')]
    public function testFnv1a(array $entry): void
    {
        $this->assertSame($entry['hash'], Fnv1a::hash($entry['input']));
        $this->assertSame($entry['hex'], Fnv1a::hex($entry['input']));
    }

    // -----------------------------------------------------------------------
    // Mulberry32
    // -----------------------------------------------------------------------

    public static function mulberry32Provider(): array
    {
        $cases = [];
        foreach (self::loadFixture('mulberry32.json') as $entry) {
            $cases['seed ' . $entry['seed']] = [$entry];
        }
        return $cases;
    }

    #[DataProvider('mulberry32Provider')]
    public function testMulberry32(array $entry): void
    {
        $m = new Mulberry32($entry['seed']);
        foreach ($entry['sequence'] as $i => $expected) {
            $float = $m->nextFloat();
            $state = $m->state();
            $this->assertSame($expected['float'], $float, "step $i float");
            $this->assertSame($expected['state'], $state, "step $i state");
        }
    }

    // -----------------------------------------------------------------------
    // Prng
    // -----------------------------------------------------------------------

    private static function prngFixture(string $method): array
    {
        static $fixture = null;
        if ($fixture === null) {
            $fixture = self::loadFixture('prng.json');
        }
        $cases = [];
        foreach ($fixture[$method] as $i => $entry) {
            $cases["#$i " . json_encode(['seed' => $entry['seed'], 'key' => $entry['key']])] = [$entry];
        }
        return $cases;
    }

    public static function prngGetValueProvider(): array
    {
        return self::prngFixture('getValue');
    }

    #[DataProvider('prngGetValueProvider')]
    public function testPrngGetValue(array $c): void
    {
        $this->assertSame($c['result'], (new Prng($c['seed']))->getValue($c['key']));
    }

    public static function prngPickProvider(): array
    {
        return self::prngFixture('pick');
    }

    #[DataProvider('prngPickProvider')]
    public function testPrngPick(array $c): void
    {
        $this->assertSame($c['result'], (new Prng($c['seed']))->pick($c['key'], $c['items']));
    }

    public static function prngWeightedPickProvider(): array
    {
        return self::prngFixture('weightedPick');
    }

    #[DataProvider('prngWeightedPickProvider')]
    public function testPrngWeightedPick(array $c): void
    {
        $this->assertSame($c['result'], (new Prng($c['seed']))->weightedPick($c['key'], $c['weights']));
    }

    public static function prngBoolProvider(): array
    {
        return self::prngFixture('bool');
    }

    #[DataProvider('prngBoolProvider')]
    public function testPrngBool(array $c): void
    {
        $this->assertSame($c['result'], (new Prng($c['seed']))->bool($c['key'], (float) $c['likelihood']));
    }

    public static function prngFloatProvider(): array
    {
        return self::prngFixture('float');
    }

    #[DataProvider('prngFloatProvider')]
    public function testPrngFloat(array $c): void
    {
        $this->assertSame((float) $c['result'], (new Prng($c['seed']))->float($c['key'], $c['range']));
    }

    public static function prngIntegerProvider(): array
    {
        return self::prngFixture('integer');
    }

    #[DataProvider('prngIntegerProvider')]
    public function testPrngInteger(array $c): void
    {
        $this->assertSame($c['result'], (new Prng($c['seed']))->integer($c['key'], $c['range']));
    }

    public static function prngShuffleProvider(): array
    {
        return self::prngFixture('shuffle');
    }

    #[DataProvider('prngShuffleProvider')]
    public function testPrngShuffle(array $c): void
    {
        $this->assertSame($c['result'], (new Prng($c['seed']))->shuffle($c['key'], $c['items']));
    }

    // -----------------------------------------------------------------------
    // Number formatting
    // -----------------------------------------------------------------------

    public static function numberProvider(): array
    {
        $cases = [];
        foreach (self::loadFixture('numbers.json') as $i => $entry) {
            $cases["#$i {$entry['output']}"] = [$entry];
        }
        return $cases;
    }

    #[DataProvider('numberProvider')]
    public function testNumberFormatting(array $entry): void
    {
        $this->assertSame($entry['output'], Number::format($entry['input']));
    }

    // -----------------------------------------------------------------------
    // Initials
    // -----------------------------------------------------------------------

    public static function initialsProvider(): array
    {
        $cases = [];
        foreach (self::loadFixture('initials.json') as $i => $entry) {
            $cases["#$i " . json_encode($entry['seed'])] = [$entry];
        }
        return $cases;
    }

    #[DataProvider('initialsProvider')]
    public function testInitials(array $entry): void
    {
        $this->assertSame($entry['result'], Initials::fromSeed($entry['seed']));
    }

    // -----------------------------------------------------------------------
    // Avatar
    // -----------------------------------------------------------------------

    public static function avatarProvider(): array
    {
        $cases = [];
        foreach (glob(self::FIXTURES_DIR . '/avatars/*.json') as $path) {
            $styleName = basename($path, '.json');
            foreach (self::loadFixture("avatars/$styleName.json") as $entry) {
                $cases["$styleName / {$entry['id']}"] = [$styleName, $entry];
            }
        }
        return $cases;
    }

    #[DataProvider('avatarProvider')]
    public function testAvatar(string $styleName, array $entry): void
    {
        static $styleCache = [];
        $styleData = $styleCache[$styleName] ??= self::loadFixture("styles/$styleName.json");

        $avatar = new Avatar($styleData, $entry['options']);
        $json = $avatar->toJSON();
        $this->assertSame($entry['svg'], $json['svg']);
        // Round-trip the resolved options through json_encode/json_decode
        // so int-vs-float typing matches the JS-generated fixture
        // (PHP `1.0` and JS `1` both become JSON `1`).
        $this->assertSame(
            $entry['resolvedOptions'],
            json_decode(json_encode($json['options'], JSON_THROW_ON_ERROR), true, flags: JSON_THROW_ON_ERROR),
        );

        // Only select cases carry a dataUri — it pins the percent-encoding
        // contract (JS encodeURIComponent) without bloating every fixture.
        if (isset($entry['dataUri'])) {
            $this->assertSame($entry['dataUri'], $avatar->toDataUri());
        }
    }

    // -----------------------------------------------------------------------
    // Color
    // -----------------------------------------------------------------------

    public static function colorProvider(): array
    {
        $fixtures = self::loadFixture('colors.json');
        $cases = [];

        foreach (['toHex', 'toRgbHex', 'parseHex', 'luminance'] as $method) {
            foreach ($fixtures[$method] as $entry) {
                $cases["$method {$entry['input']}"] = [$method, [$entry['input']], $entry['result']];
            }
        }

        foreach ($fixtures['sortByContrast'] as $i => $entry) {
            $cases["sortByContrast #$i"] = [
                'sortByContrast',
                [$entry['candidates'], $entry['refColor']],
                $entry['result'],
            ];
        }

        foreach ($fixtures['filterNotEqualTo'] as $i => $entry) {
            $cases["filterNotEqualTo #$i"] = [
                'filterNotEqualTo',
                [$entry['candidates'], $entry['excluded']],
                $entry['result'],
            ];
        }

        return $cases;
    }

    /**
     * @param list<mixed> $args
     */
    #[DataProvider('colorProvider')]
    public function testColor(string $method, array $args, mixed $expected): void
    {
        if ($method === 'luminance') {
            // JSON serializes whole-number luminance (0, 1) as int.
            $expected = (float) $expected;
        }

        $this->assertSame($expected, Color::$method(...$args));
    }

    // -----------------------------------------------------------------------
    // Validation
    // -----------------------------------------------------------------------
    //
    // Only the accept/reject outcome is shared across languages — error
    // messages are language-specific. The circular color reference cases
    // additionally pin the reported resolution chain.

    public static function styleValidationProvider(): array
    {
        $cases = [];
        foreach (self::loadFixture('validation.json')['styles'] as $entry) {
            $cases[$entry['id']] = [$entry];
        }
        return $cases;
    }

    #[DataProvider('styleValidationProvider')]
    public function testStyleValidation(array $entry): void
    {
        if ($entry['valid']) {
            $this->assertInstanceOf(Style::class, new Style($entry['definition']));
        } else {
            $this->expectException(ValidationError::class);
            new Style($entry['definition']);
        }
    }

    public static function optionsValidationProvider(): array
    {
        $cases = [];
        foreach (self::loadFixture('validation.json')['options'] as $entry) {
            $cases[$entry['id']] = [$entry];
        }
        return $cases;
    }

    #[DataProvider('optionsValidationProvider')]
    public function testOptionsValidation(array $entry): void
    {
        static $minimalStyle = null;
        $minimalStyle ??= array_values(array_filter(
            self::loadFixture('validation.json')['styles'],
            static fn (array $e) => $e['id'] === 'minimal',
        ))[0]['definition'];

        if ($entry['valid']) {
            $this->assertInstanceOf(Avatar::class, new Avatar($minimalStyle, $entry['options']));
        } else {
            $this->expectException(ValidationError::class);
            new Avatar($minimalStyle, $entry['options']);
        }
    }

    public static function circularColorProvider(): array
    {
        $cases = [];
        foreach (self::loadFixture('validation.json')['circularColors'] as $entry) {
            $cases[$entry['id']] = [$entry];
        }
        return $cases;
    }

    #[DataProvider('circularColorProvider')]
    public function testCircularColorReference(array $entry): void
    {
        try {
            new Avatar($entry['style'], $entry['options']);
            $this->fail('expected a circular color reference error');
        } catch (CircularColorReferenceError $e) {
            $this->assertSame($entry['chain'], $e->chain);
        }
    }

    // -----------------------------------------------------------------------
    // OptionsDescriptor
    // -----------------------------------------------------------------------

    public static function descriptorProvider(): array
    {
        $cases = [];
        foreach (glob(self::FIXTURES_DIR . '/descriptors/*.json') as $path) {
            $styleName = basename($path, '.json');
            $cases[$styleName] = [$styleName];
        }
        return $cases;
    }

    #[DataProvider('descriptorProvider')]
    public function testOptionsDescriptor(string $styleName): void
    {
        $style = new Style(self::loadFixture("styles/$styleName.json"));
        $descriptor = (new OptionsDescriptor($style))->toJSON();

        // Same round-trip as the avatar test: pins key order and int-vs-float
        // typing against the JS-generated fixture.
        $this->assertSame(
            self::loadFixture("descriptors/$styleName.json"),
            json_decode(json_encode($descriptor, JSON_THROW_ON_ERROR), true, flags: JSON_THROW_ON_ERROR),
        );
    }
}
